modules/build/src/main/scala/scala/build/internal/ObjectCodeWrapper.scala

```diff
@@ -4,66 +4,92 @@ package scala.build.internal
-                                    |}
-                                    |""".stripMargin)
+    val mainObjectName     = CodeWrapper.mainClassObject(indexedWrapperName).backticked
+    val wrapperObjectName  = indexedWrapperName.backticked
+    val aliasedWrapperName = wrapperObjectName + "$$alias"
+
+    val wrapperReference = if (wrapperObjectName == "main")
+      s"$aliasedWrapperName.alias" // https://github.com/VirtusLab/scala-cli/issues/314
+    else wrapperObjectName
 
     val packageDirective =
       if (pkgName.isEmpty) "" else s"package ${AmmUtil.encodeScalaSourcePath(pkgName)}" + "\n"
 
-    val aliasObject =
-      if (name == "main_sc")
+    val mainObject = AmmUtil.normalizeNewlines(
+      s"""object $mainObjectName {
+         |  private var args$$opt0 = Option.empty[Array[String]]
+         |  def args$$set(args: Array[String]): Unit = {
+         |    args$$opt0 = Some(args)
+         |  }
+         |  def args$$opt: Option[Array[String]] = args$$opt0
+         |  def args$$: Array[String] = args$$opt.getOrElse {
+         |    sys.error("No arguments passed to this script")
+         |  }
+         |  def main(args: Array[String]): Unit = {
+         |    args$$set(args)
+         |    val _ = $wrapperReference${if useDelayedInit then ".run()" else ".hashCode()"}
+         |  }
+         |}
+         |""".stripMargin
+    )
+
+    val aliasObject = if (wrapperObjectName == "main")
+      AmmUtil.normalizeNewlines(
         s"""object $aliasedWrapperName {
-           |  val alias = ${indexedWrapperName.backticked}
-           |}""".stripMargin
-      else ""
+           |  val alias = $wrapperObjectName
+           |}
+           |""".stripMargin
+      )
+    else ""
 
-    // indentation is important in the generated code, so we don't want scalafmt to touch that
-    // format: off
-    val top = AmmUtil.normalizeNewlines(s"""
-$packageDirective
+    val extendsClause = if useDelayedInit then "extends scala.cli.build.ScalaCliApp " else ""
 
-
-object ${indexedWrapperName.backticked} {
-def args = $name.args$$
-def scriptPath = \"\"\"$scriptPath\"\"\"
-""")
-    val bottom = AmmUtil.normalizeNewlines(s"""
-$extraCode
-}
-$aliasObject
-$mainObjectCode
-""")
-    // format: on
+    val top = AmmUtil.normalizeNewlines(
+      s"""$packageDirective
+         |
+         |object $wrapperObjectName $extendsClause{
+         |def args = $mainObjectName.args$$
+         |def scriptPath = \"\"\"$scriptPath\"\"\"
+         |""".stripMargin
+    )
+    val bottom = AmmUtil.normalizeNewlines(
+      s"""
+         |$extraCode
+         |}
+         |
+         |$mainObject
+         |
+         |$aliasObject
+         |""".stripMargin
+    )
 
     (top, bottom)
   }
+
+  override def additionalSourceCode: Option[(os.RelPath, String)] = Option.when(useDelayedInit)(
+    os.RelPath("delayed-init-wrapper.scala") ->
+      // This is copied from Scala 2 implementation of App trait, but with no main method and no argument handling
```

Comment:
might want to include the commit hash of the copied code from the Scala 2 repo, for future context (even if the code isn't subject to much changes between versions, as I'd presume)